File: generated-code/ruby/codecraft/TcpReadWrite/tcp_stream.rb
require 'stringio'
require 'socket'
require_relative 'stream'

class TcpStream < Stream
    def initialize(host, port)
        @socket = TCPSocket.open(host, port)
        @socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
        @read_buffer = StringIO.new('', 'a+b')
        @write_buffer = ''
    end

    def close
        @socket.close
    end

    def read_bytes(byte_count)
        @read_buffer = StringIO.new('', 'a+b') unless @read_buffer.is_a? StringIO

        data = @read_buffer.read(byte_count) || ''

        while data.length < byte_count
            @read_buffer = StringIO.new(@socket.recv(102_400), 'a+b')
            data << @read_buffer.read(byte_count - data.length)
        end

        data
    end

    def write_bytes(data)
        @write_buffer << data
    end

    def flush
        @socket.write(@write_buffer)
        @write_buffer = ''
        @read_buffer = ''
        @socket.flush
    end
end

File: generated-code/typescript/codecraft/TcpReadWrite/src/tcp-stream.ts
import { Stream } from "./stream";
import { Socket } from "net";

export class TcpStream extends Stream {
    private socket: Socket;
    private data: Buffer;
    private needAmount: number | null;
    private resolve: any;
    constructor(host: string, port: number) {
        super();
        const _this = this;
        this.socket = new Socket().setNoDelay(true);
        this.socket
            .setNoDelay(true)
            .on('error', (error) => {
                console.error('Socket error: ' + error.message);
                process.exit(1);
            });
        new Promise(function (resolve) {
            _this.socket.connect(port, host, () => resolve(undefined));
        });
        this.data = Buffer.alloc(0);
        this.needAmount = null;
        this.resolve = null;
        this.socket.on('data', function (data) { _this.dataHandler(data) });
    }
    destroy() {
        this.socket.destroy();
    }
    dataHandler(data: Buffer) {
        this.data = Buffer.concat([this.data, data]);
        this.update();
    }
    update() {
        if (this.needAmount === null || this.needAmount > this.data.length) {
            return;
        }
        const data = this.data.slice(0, this.needAmount);
        this.data = this.data.slice(this.needAmount);
        this.needAmount = null;
        this.resolve(data);
        this.update();
    }
    async read(byteCount: number): Promise<Buffer> {
        const _this = this;
        return await new Promise<Buffer>(function (resolve, reject) {
            _this.needAmount = byteCount;
            _this.resolve = resolve;
            _this.update();
        });
    }
    async write(data: Buffer): Promise<void> {
        const _this = this;
        return await new Promise<void>(function (resolve, reject) {
            _this.socket.write(data, 'utf8', function (error) {
                if (error) {
                    return reject(error);
                }
                resolve(undefined);
            });
        });
    }
    async flush() { }
}